librz/analysis/var_global.c

```diff
@@ -0,0 +1,370 @@
+
+#include <rz_analysis.h>
+#include <rz_list.h>
+
+/**
+ * \brief Create a new instance of global variable
+ * 
+ * \param name variable name
+ * \param addr variable address
+ * \param comment variable comment
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_OWN RzAnalysisVarGlobal *rz_analysis_var_global_new(RZ_NONNULL const char *name, ut64 addr) {
+	rz_return_val_if_fail(name, NULL);
+	RzAnalysisVarGlobal *glob = RZ_NEW0(RzAnalysisVarGlobal);
+	if (!glob) {
+		return NULL;
+	}
+	glob->name = strdup(name);
+	glob->addr = addr;
+	return glob;
+}
+
+typedef struct {
+	RBNode rb;
+	ut64 addr;
+	ut64 size;
+} GlobalVarNode;
+
+int global_var_node_cmp(const void *incoming, const RBNode *in_tree, void *user) {
+	ut64 ia = *(ut64 *)incoming;
+	ut64 ta = container_of(in_tree, const GlobalVarNode, rb)->addr;
+	if (ia < ta) {
+		return -1;
+	} else if (ia > ta) {
+		return 1;
+	}
+	return 0;
+}
+
+/**
+ * \brief Add the global variable into hashtable
+ * 
+ * \param analysis RzAnalysis
+ * \param global_var Global variable instance
+ * \return true if succeed
+ */
+RZ_API RZ_OWN bool rz_analysis_var_global_add(RzAnalysis *analysis, RzAnalysisVarGlobal *global_var) {
+	rz_return_val_if_fail(analysis && global_var, false);
+	if (rz_analysis_var_global_get_byaddr(analysis, global_var->addr)) {
+		RZ_LOG_ERROR("Global variable %s at 0x%" PFMT64x " already exists!\n", global_var->name, global_var->addr);
+		return false;
+	} else if (ht_pp_find(analysis->ht_global_var, global_var->name, NULL)) {
+		RZ_LOG_ERROR("Global variable %s already exists!\n", global_var->name);
+		return false;
+	}
+	if (!ht_pp_insert(analysis->ht_global_var, global_var->name, global_var)) {
+		return false;
+	}
+	GlobalVarNode *node = RZ_NEW0(GlobalVarNode);
+	node->addr = global_var->addr;
+	node->size = 0;
+	if (global_var->type) {
+		node->size = rz_type_db_get_bitsize(analysis->typedb, global_var->type) / 8;
+	}
+	if (!rz_rbtree_aug_insert(&analysis->global_var_tree, &node->addr, &node->rb, global_var_node_cmp, NULL, NULL)) {
+		return false;
+	}
+	return true;
+}
+
+/**
+ * \brief Free the global variable instance
+ * 
+ * \param glob Global variable instance
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_free(RzAnalysisVarGlobal *glob) {
+	if (!glob) {
+		return;
+	}
+
+	RZ_FREE(glob->name);
+	rz_type_free(glob->type);
+	rz_vector_fini(&glob->constraints);
+	RZ_FREE(glob);
+}
+
+/**
+ * \brief Delete and free the global variable by its name
+ * 
+ * \param analysis RzAnalysis
+ * \param name Global Variable name
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_delete_byname(RzAnalysis *analysis, const char *name) {
+	rz_return_val_if_fail(analysis && name, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byname(analysis, name);
+	if (!glob) {
+		return false;
+	}
+	rz_rbtree_delete(&analysis->global_var_tree, &glob->addr, global_var_node_cmp, NULL, NULL, NULL);
+	return ht_pp_delete(analysis->ht_global_var, name);
```

Comment:
I suggest to delete things in the same way you are doing in rz_analysis_var_global_delete_byaddr.

Reply:
Why? This way it's shorter and just one call.

Reply:
Yeah, we've got the global variable here, no need to call `rz_analysis_var_global_delete_byaddr` to get again.